Make discussion links into buttons for a11y

andy-armstrong opened the pull request:
### Description
 
[TNL-4640](https://openedx.atlassian.net/browse/TNL-4640)

To improve the accessibility of discussions, this change converts anchor tags to be buttons if they are only used for actions. 

### Sandbox
- [x] https://andy-armstrong.sandbox.edx.org

### Testing
 - [ ] manual testing of the discussion board, inline discussions, and team discussions

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [x] Code review: @bjacobel 
- [x] Code review: @dianakhuang 
- [x] Doc Review: @catong 
- [x] UX review: @chris-mike 
- [ ] Accessibility review: @clrux 
- [x] Product review: @marcotuts 

FYI: Tag anyone who might be interested in this PR here.

### Post-review
- [ ] Rebase and squash commits

bjacobel commented on lms/static/sass/discussion/elements/_actions.scss:
```diff
@@ -141,18 +141,15 @@
     .action-label {
       @extend %t-copy-sub2;
-      display: inline-block;
+      display: none;
```
This may be an a11y issue - previously this was `opacity: 0` and now it's `display: none`, IIRC screenreaders won't read `display: none` elements?

(the real fix would be to use the `.sr-only` class here, as it seems like that's what this was trying to achieve)

andy-armstrong replied:
The opacity was an issue because it was rendering white on white, but the screen reader still read it. I made it ``display: none`` because I don't want anyone to see it, screen readers included. The screen reader will read it when it is given the focus.

@clrux what's your opinion? Should the hiding be a visual-only thing, and just have the screenreader always read it?

clrux replied:
I advocate for an equal experience.

If you don't want anyone to see it, why include it in the markup?

clrux replied:
So this only appears on focus?

andy-armstrong replied:
@clrux Yep, this appears only on hover and focus. The previous implementation hid the button by making its opacity 0, so I changed it to just use display: none.

clrux replied:
Cool. I think this should work. If it's focused, then it should be announced.

bjacobel commented:
LGTM aside from one a11y issue I might have found.

dianakhuang commented:
Yay! 👍

chris-mike commented:
I don't see an obvious issues. 👍

clrux commented on lms/static/sass/discussion/utilities/_shame.scss:
```diff
@@ -7,8 +7,12 @@
+
+      .fa {
+        @include margin-right(10px);
+      }
```
https://s-media-cache-ak0.pinimg.com/236x/3d/8e/34/3d8e34678a4851b374f882f8f5d0570f.jpg

clrux commented on lms/static/sass/discussion/utilities/_variables-v1.scss:
```diff
@@ -15,7 +15,7 @@ $forum-color-community-ta: $green-d1 !default;
 $forum-color-border: $gray-l3 !default;
 $forum-color-error: $red !default;
-$forum-color-hover-thread: $gray-d3 !default;
+$forum-color-hover-thread: #f6f6f6 !default;
```
Is this to fill a missing graygap?

andy-armstrong replied:
This is in the non-Pattern Library version, so there probably is a variable I could have used. It isn't affected by the set of grays in the Pattern Library.

clrux commented on lms/static/sass/discussion/views/_thread.scss:
```diff
@@ -257,17 +257,17 @@ body.discussion {
             }
 
-            a {
+            .btn-link {
```
Sweet!

clrux commented on common/test/acceptance/tests/discussion/test_discussion.py:
```diff
@@ -211,7 +211,6 @@ def test_page_accessibility(self):
             "ignore": [
                 'section',  # TODO: AC-491
                 'color-contrast',  # TNL-4635
-                'link-href',  # TNL-4636
```
SWEET!

clrux commented on common/static/common/templates/discussion/search-alert.underscore:
```diff
@@ -1,10 +1,12 @@
     <div class="search-alert-controls">
-      <a href="#" class="dismiss control control-dismiss"><span class="icon fa fa-remove" aria-hidden="true"></span>
-      <span class="sr"><%- gettext("Close") %></span></a>
+        <button class="btn-link dismiss control control-dismiss">
```
You could also use `title` here instead of the `sr` if you preferred.

andy-armstrong replied:
Fixed

clrux commented on common/static/common/templates/discussion/response-comment-edit.underscore:
```diff
@@ -1,9 +1,9 @@
-  </div>
-  <input type="submit" id="edit-comment-submit" class="post-update" value="<%- gettext("Update comment") %>">
-  <a href="#" class="post-cancel"><%- gettext("Cancel") %></a>
+    <h1><%- gettext("Editing comment") %></h1>
```
Does it make sense to use `.hd-1` here to utilize UXPL headings when we bring them in?

andy-armstrong replied:
This is used in a non-pattern library context as well. Would that work in both? I'd rather have a separate PR to clean up the headings.

clrux replied:
Yep! If it's not defined (which it isn't outside of UXPL things) then it won't do anything.

clrux replied:
A while back I created a PR that brought in the headings to LMS and CMS, so using this class would make this heading look like those defined in the UXPL. A later PR for cleanup is more than fine.

clrux commented on common/static/common/templates/discussion/forum-actions.underscore:
```diff
@@ -3,10 +3,10 @@
             <div class="more-wrapper">
-                <a href="javascript:void(0)" class="action-button action-more" role="button" aria-haspopup="true" aria-controls="action-menu-<%= contentId %>">
-                    <span class="action-label"><%- gettext("More") %></span>
+                <button class="btn-link action-button action-more" aria-label="<%- gettext('More') %>" aria-haspopup="true" aria-controls="action-menu-<%= contentType %>">
```
Not sure the `aria-label` is  necessary since there's text within the button.

andy-armstrong replied:
I was getting an error from the a11y acceptance tests without it. Any ideas why that would be?

clrux replied:
Hm, what was the error?

clrux commented on common/static/common/templates/discussion/forum-action-vote.underscore:
```diff
@@ -2,7 +2,7 @@
       <span class="vote-count"></span>
     </span>
-    <a href="#" class="action-button action-vote" role="checkbox" aria-checked="false">
+    <button class="btn-link action-button action-vote" role="checkbox" aria-checked="false">
```
Is there JavaScript keeping tabs of the new role here? Meaning, when this is pressed, does `aria-checked` change its value accordingly? Any reason not to use an actual checkbox input type here?

clrux replied:
Also, is this an action, or a toggle?

andy-armstrong replied:
Good question. It is kind-of a toggle, but it seems more like an action to me. Can I just remove the role and aria-checked attributes?

andy-armstrong replied:
In fact, it is a toggle. When it is clicked, this is the JavaScript that is updating it:
https://github.com/edx/edx-platform/blob/master/common/static/common/js/discussion/views/discussion_content_view.js#L247
```javascript
        DiscussionContentShowView.prototype.updateButtonState = function(selector, checked) {
            var $button;
            $button = this.$(selector);
            $button.toggleClass('is-checked', checked);
            return $button.attr('aria-checked', checked);
        };
```

andy-armstrong replied:
@clrux and I discussed this and we're going to leave it as is for now. I've made a JIRA ticket to add toggle buttons to the pattern library:

https://openedx.atlassian.net/browse/FEDX-238

clrux commented on common/static/common/templates/discussion/forum-action-report.underscore:
```diff
@@ -1,5 +1,5 @@
 <li class="actions-item">
-    <a href="javascript:void(0)" class="action-list-item action-report" role="checkbox" aria-checked="false">
+    <button class="btn-link action-list-item action-report" role="checkbox" aria-checked="false">
```
Is there JavaScript keeping tabs of the new role here? Meaning, when this is pressed, does `aria-checked` change its value accordingly? Any reason not to use an actual checkbox input type here?

clrux replied:
Also, is this an action, or a toggle?

clrux commented on common/static/common/templates/discussion/forum-action-pin.underscore:
```diff
@@ -1,5 +1,5 @@
 <li class="actions-item">
-    <a href="javascript:void(0)" class="action-list-item action-pin" role="checkbox" aria-checked="false">
+    <button class="btn-link action-list-item action-pin" role="checkbox" aria-checked="false">
```
Is there JavaScript keeping tabs of the new role here? Meaning, when this is pressed, does `aria-checked` change its value accordingly? Any reason not to use an actual checkbox input type here?

clrux replied:
Also, is this an action, or a toggle?

clrux commented on common/static/common/templates/discussion/forum-action-follow.underscore:
```diff
@@ -1,10 +1,10 @@
 <li class="actions-item">
-    <a href="javascript:void(0)" class="action-button action-follow" role="checkbox" aria-checked="false">
+    <button class="btn-link action-button action-follow" role="checkbox" aria-checked="false">
```
Is there JavaScript keeping tabs of the new role here? Meaning, when this is pressed, does `aria-checked` change its value accordingly? Any reason not to use an actual checkbox input type here?

clrux replied:
Also, is this an action, or a toggle?

clrux commented on common/static/common/templates/discussion/forum-action-endorse.underscore:
```diff
@@ -1,10 +1,10 @@
 <li class="actions-item">
-    <a href="javascript:void(0)" class="action-button action-endorse" role="checkbox" aria-checked="false">
+    <button class="btn-link action-button action-endorse" role="checkbox" aria-checked="false">
```
Is there JavaScript keeping tabs of the new role here? Meaning, when this is pressed, does `aria-checked` change its value accordingly? Any reason not to use an actual checkbox input type here?

clrux replied:
Also, is this an action, or a toggle?

clrux commented on common/static/common/templates/discussion/forum-action-edit.underscore:
```diff
@@ -1,6 +1,6 @@
 <li class="actions-item">
-    <a href="javascript:void(0)" class="action-list-item action-edit" role="button">
+    <button class="btn-link action-list-item action-edit" role="button">
```
Don't need the role here. It's already a button! :)

andy-armstrong replied:
Fixed

clrux commented on common/static/common/templates/discussion/forum-action-delete.underscore:
```diff
@@ -1,6 +1,6 @@
 <li class="actions-item">
-    <a href="javascript:void(0)" class="action-list-item action-delete" role="button">
+    <button class="btn-link action-list-item action-delete" role="button">
```
Don't need the role here. It's already a button! :)

andy-armstrong replied:
Fixed

clrux commented on common/static/common/templates/discussion/forum-action-close.underscore:
```diff
@@ -1,5 +1,5 @@
 <li class="actions-item">
-    <a href="javascript:void(0)" class="action-list-item action-close" role="checkbox" aria-checked="false">
+    <button class="btn-link action-list-item action-close" role="checkbox" aria-checked="false">
```
Is there JavaScript keeping tabs of the new role here? Meaning, when this is pressed, does `aria-checked` change its value accordingly? Any reason not to use an actual checkbox input type here?

clrux replied:
Also, is this an action or a toggle?

clrux commented on common/static/common/templates/discussion/forum-action-answer.underscore:
```diff
@@ -1,10 +1,10 @@
 <li class="actions-item">
-    <a href="javascript:void(0)" class="action-button action-answer" role="checkbox" aria-checked="false">
+    <button class="btn action-button action-answer" role="checkbox" aria-checked="false">
```
Is there JavaScript keeping tabs of the new role here? Meaning, when this is pressed, does `aria-checked` change its value accordingly? Any reason not to use an actual checkbox input type here?

clrux replied:
Also, is this an action or a toggle?

catong commented:
@andy-armstrong No doc review needed for this PR. Thanks!

andy-armstrong commented:
@clrux I've addressed your feedback and squashed ready for merging. They were trivial changes so I'll merge once I get a green build if that's okay with you.

andy-armstrong commented:
jenkins run bokchoy
